vnext/Microsoft.ReactNative/RedBox.cpp

```diff
@@ -346,26 +347,24 @@ struct RedBox : public std::enable_shared_from_this<RedBox> {
 /*
  * This class is implemented such that the methods on IRedBoxHandler are thread safe.
  */
-struct RedBoxHandler : public std::enable_shared_from_this<RedBoxHandler>, IRedBoxHandler {
-  RedBoxHandler(
-      Mso::WeakPtr<Mso::React::IReactHost> weakReactHost,
-      std::shared_ptr<facebook::react::MessageQueueThread> uiMessageQueue) noexcept
-      : m_weakReactHost(std::move(weakReactHost)), m_wkUIMessageQueue(std::move(uiMessageQueue)) {}
+struct DefaultRedBoxHandler : public std::enable_shared_from_this<DefaultRedBoxHandler>, IRedBoxHandler {
+  DefaultRedBoxHandler(Mso::WeakPtr<Mso::React::IReactHost> &&weakReactHost) noexcept
+      : m_weakReactHost(std::move(weakReactHost)),
+        m_uiDispatcher(winrt::Windows::ApplicationModel::Core::CoreApplication::MainView().CoreWindow().Dispatcher()) {}
```

Comment:
please don't add references to core window or coredispatcher. Use DispatcherQueue

Reply:
I dont see a way from DispatcherQueue to get the UIQueue?  I was trying to avoid having to require the UI dispatcher queue as one of the ctor parameters on the DefaultRedBox impl. 

Side note, does Xaml::Window work on islands?  -- since redbox is currently using that too.

Reply:
you'd do: auto dispatcher = winrt::Windows::System::DispatcherQueue::GetForCurrentThread(); from the UI thread, and pass it as a lambda capture whenever you need to post back to the ui thread.

Reply:
right, so that will affect the public API, because the RN instance can be created from a background thread, so we can't use GetForCurrentThread.  So I"ll need them to provide a UI dispatchqueue to the ctor of the redbox handler.  -- Which is fine if I cant use the corewindow API to get the UI thread.

Reply:
This pattern is in a few places. E.g. Mso code will often ask for the current UI Queue which uses the CoreWindow under the hood. 

Single instance multiple UI thread is kind of scary for native modules that have a UI-thread-affined component. I get the feeling that even on XAML islands we'll have an instance global UI dispatcher on the property bag instead of something per-view-tree.

Reply:
I'm sure @vmoroz will have some opinions on this

Reply:
Per our discussion today, I will add a new ABI interface for dispatch queues and we will be able to pass it to RNH. Then we could retrieve it here.

---
In reply to: [419833368](https://github.com/microsoft/react-native-windows/pull/4786#discussion_r419833368) [](ancestors = 419833368)

Reply:
For the time being you could use the DispatchQueue::MainUIQueue(). It is currently based on the CoreDispatcher, but I will switch it to the DispatcherQueue as soon as possible.

---
In reply to: [420454784](https://github.com/microsoft/react-native-windows/pull/4786#discussion_r420454784) [](ancestors = 420454784,419833368)